transport-classes-io_uring/src/main/java/io/netty/channel/uring/IOUringIoHandler.java

```diff
@@ -0,0 +1,432 @@
+                public void handle(int res, int flags, int id, byte op, short data) {
+                    if (id == EVENTFD_ID) {
+                        eventFdDrained = true;
+                    }
+                    IOUringIoHandler.this.handle(res, flags, id, op, data);
+                }
+            }
+            final DrainFdEventCallback handler = new DrainFdEventCallback();
+            completionQueue.process(handler);
+            while (!handler.eventFdDrained) {
+                submissionQueue.submitAndWait();
+                completionQueue.process(handler);
+            }
+        }
+        // We've consumed any pending eventfd read and `eventfdAsyncNotify` should never
+        // transition back to false, thus we should never have any more events written.
+        // So, if we have a read event pending, we can cancel it.
+        if (eventfdReadSubmitted != 0) {
+            submissionQueue.addCancel(eventfd.intValue(), eventfdReadSubmitted, EVENTFD_ID);
+            eventfdReadSubmitted = 0;
+            submissionQueue.submit();
+        }
+    }
+
+    private void completeRingClose() {
+        if (closeCompleted) {
+            // already done.
+            return;
+        }
+        closeCompleted = true;
+        ringBuffer.close();
+        try {
+            eventfd.close();
+        } catch (IOException e) {
+            logger.warn("Failed to close eventfd", e);
+        }
+        PlatformDependent.freeMemory(eventfdReadBuf);
+    }
+
+    @Override
+    public IoRegistration register(IoEventLoop eventLoop, IoHandle handle) throws Exception {
+        IOUringIoHandle ioHandle = cast(handle);
+        if (shuttingDown) {
+            throw new RejectedExecutionException("IoEventLoop is shutting down");
+        }
+        DefaultIoUringIoRegistration registration = new DefaultIoUringIoRegistration(eventLoop, ioHandle);
+        for (;;) {
+            int id = nextRegistrationId();
+            DefaultIoUringIoRegistration old = registrations.put(id, registration);
+            if (old != null) {
+                assert old.handle != registration.handle;
+                registrations.put(id, old);
+            } else {
+                registration.id = id;
+                break;
+            }
+        }
+
+        ringBuffer.ioUringSubmissionQueue().incrementHandledFds();
+        return registration;
+    }
+
+    private int nextRegistrationId() {
+        int id;
+        do {
+            id = nextRegistrationId++;
+        } while (id == RINGFD_ID || id == EVENTFD_ID);
+        return id;
+    }
+
+    private final class DefaultIoUringIoRegistration extends AtomicBoolean implements IOUringIoRegistration {
+        private final IoEventLoop eventLoop;
+        private final IOUringIoEvent event = new IOUringIoEvent(0, 0, (byte) 0, 0, (short) 0);
+        final IOUringIoHandle handle;
+
+        private boolean removeLater;
+        private int outstandingCompletions;
+
+        int id;
```

Comment:
I see you are updating it directly, and is fine, but is still pretty stealthy without a method to assign it

Reply:
imho it does not matter. But if others want to have a method we can also add one